services-geocoding/src/main/java/com/mapbox/api/geocoding/v5/models/GeocodingTypeAdapter.java

```diff
@@ -0,0 +1,120 @@
+package com.mapbox.api.geocoding.v5.models;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.google.gson.Gson;
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ *
+ * @since 4.9.0
+ */
+abstract class GeocodingTypeAdapter<T> extends TypeAdapter<T> {
+
+  volatile TypeAdapter<String> stringAdapter;
+  final Gson gson;
+
+  GeocodingTypeAdapter(Gson gson) {
+    this.gson = gson;
+  }
+
+  void writeString(@NonNull String name, String value,
+                           @NonNull JsonWriter jsonWriter) throws IOException {
+    jsonWriter.name(name);
+    if (value == null) {
+      jsonWriter.nullValue();
+    } else {
+      TypeAdapter<String> stringAdapter = this.stringAdapter;
+      if (stringAdapter == null) {
+        this.stringAdapter = stringAdapter;
+        stringAdapter = gson.getAdapter(String.class);
```

Comment:
Looks like the adapter always uses `String.class`, does it have to be lazily initialized in each method instead once in the class field?

Reply:
This is the part that I was not sure about. 
The code is extracted from generated AutoValue.
There is an instance variable: stringAdapter which will be Lazily initialized.
I am not sure why there is a need to keep a local copy of `this. stringAdapter` 

Decided to keep it as is. I think I should address that part separately.